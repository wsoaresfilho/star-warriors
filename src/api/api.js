// Google Search API Key (gsak) has to be added down here
const gsak = '';
const ckk = '002028002876040441306:7gjv47i0lhn';
const apiBase = 'https://swapi.co/api';
const googleImgApi = `https://www.googleapis.com/customsearch/v1?key=${gsak}&cx=${ckk}&searchType=image&imgSize=large&imgType=photo&num=1&q=`;

export const getDataFromApi = async url => {
    const response = await fetch(url || `${apiBase}/people/`);
    const myJson = await response.json();
    return myJson;
};

export const getImageByTermFake = () => {
    const response = new Promise(resolve => {
        setTimeout(() => {
            resolve({
                items: [
                    {
                        link: 'logo-black.svg',
                    },
                ],
            });
        }, 100);
    });
    return response;
};

export const getImageByTerm = async imageTerm => {
    const response = await fetch(`${googleImgApi}${encodeURI(imageTerm)}`);
    if (response.status >= 400) {
        return getImageByTermFake();
    }
    const myJson = await response.json();
    return myJson;
};
