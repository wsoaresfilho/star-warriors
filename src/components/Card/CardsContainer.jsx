import React, { PureComponent } from 'react';
import Card from './Card';
import { getDataFromApi, getImageByTerm } from '../../api/api';
import getIntFromString from '../../utils/helpers';

class CardsContainer extends PureComponent {
    constructor(props) {
        super(props);

        this.state = {
            cards: [],
            isFetching: false,
            next: null,
            addedCardsNum: 0,
            starshipsMap: new Map(),
        };
    }

    componentDidMount() {
        this.fetchCards();
    }

    componentDidUpdate() {
        const { cards, isFetching } = this.state;
        if (cards.length > 0 && !isFetching) {
            this.startLazyLoadingCards();
        }
    }

    getLoadingElement() {
        const { isFetching } = this.state;
        return (
            <div className='loading'>
                {isFetching && <img src='loading.gif' alt='Loading' />}
            </div>
        );
    }

    async fetchCards() {
        this.setState({ isFetching: true });

        const { next: link, cards, starshipsMap } = this.state;
        const people = await getDataFromApi(link);
        const { results, next } = people;

        if (results.length > 0) {
            const updatedResults = await Promise.all(
                results.map(async res => {
                    const imageResp = await getImageByTerm(res.name);
                    const imageUrl =
                        imageResp.items && imageResp.items.length > 0
                            ? imageResp.items[0].link
                            : '';
                    const { starships } = res;

                    const starshipsDetails = await Promise.all(
                        starships.map(async sts => {
                            const starShipId = getIntFromString(sts);
                            const stsMap = starshipsMap.has(starShipId);
                            if (stsMap) {
                                return starshipsMap.get(starShipId);
                            }

                            const newStarship = await getDataFromApi(sts);
                            this.setState({
                                starshipsMap: starshipsMap.set(
                                    starShipId,
                                    newStarship
                                ),
                            });
                            return newStarship;
                        })
                    );

                    const updatedRes = {
                        ...res,
                        imgUrl: imageUrl,
                        starshipsDetails,
                    };
                    return updatedRes;
                })
            );

            this.setState({
                cards: cards.concat(updatedResults),
                isFetching: false,
                next,
                addedCardsNum: results.length,
            });
        }
    }

    startLazyLoadingCards() {
        const { next, addedCardsNum } = this.state;
        const elements = [].slice.call(
            document.querySelectorAll('.cards-container img')
        );

        if ('IntersectionObserver' in window) {
            const lazyImageObserver = new IntersectionObserver(entries => {
                entries.forEach(async entry => {
                    if (entry.isIntersecting) {
                        const lazyImage = entry.target;
                        lazyImageObserver.unobserve(lazyImage);

                        if (next) {
                            this.fetchCards();
                        }
                    }
                });
            });

            elements.slice(-addedCardsNum).forEach((lazyImage, index) => {
                if ((index + 1) % 7 === 0) {
                    lazyImageObserver.observe(lazyImage);
                }
            });
        }
    }

    render() {
        const { cards } = this.state;
        return (
            <div className='cards-main'>
                <h1 className='cards-container__title'>Star Wars Characters</h1>
                <div className='cards-container'>
                    {cards.map(card => {
                        const props = card;
                        return <Card key={card.name} {...props} />;
                    })}
                    {this.getLoadingElement()}
                </div>
            </div>
        );
    }
}

export default CardsContainer;
