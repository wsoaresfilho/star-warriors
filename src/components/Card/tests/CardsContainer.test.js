import React from 'react';
import { mount } from 'enzyme';
import CardsContainer from '../CardsContainer';
import * as api from '../../../api/api';
import { peopleMock, imageMock, starshipsMock } from './mocks';

describe('CardsContainer component', () => {
    let wrapper;

    beforeEach(() => {
        api.getDataFromApi = jest.fn(
            url =>
                new Promise(resolve => {
                    if (url && url.includes('starships')) {
                        resolve(starshipsMock);
                    } else {
                        resolve(peopleMock);
                    }
                })
        );
        api.getImageByTerm = jest.fn(
            () =>
                new Promise(resolve => {
                    resolve(imageMock);
                })
        );

        wrapper = mount(<CardsContainer />);
    });

    describe('when initializes should render', () => {
        it('cards-main element', done => {
            setImmediate(() => {
                wrapper.update();
                expect(wrapper.find('.cards-main').length).toBe(1);
                done();
            });
        });

        it('cards-container title', done => {
            setImmediate(() => {
                wrapper.update();
                expect(wrapper.find('.cards-container__title').text()).toBe(
                    'Star Wars Characters'
                );
                done();
            });
        });

        it('Cards', done => {
            setImmediate(() => {
                wrapper.update();
                expect(wrapper.find('Card').length).toBe(4);
                done();
            });
        });

        it('snapshot correctly', done => {
            setImmediate(() => {
                wrapper.update();
                expect(wrapper).toMatchSnapshot();
                done();
            });
        });
    });

    describe('when fetching should render', () => {
        it('loading element', done => {
            setImmediate(() => {
                wrapper.setState({ isFetching: true });
                wrapper.update();
                expect(wrapper.find('.loading').length).toBe(1);
                done();
            });
        });

        it('snapshot correctly', done => {
            setImmediate(() => {
                wrapper.update();
                expect(wrapper).toMatchSnapshot();
                done();
            });
        });
    });
});
