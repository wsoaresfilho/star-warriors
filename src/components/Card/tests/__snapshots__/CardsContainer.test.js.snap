// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardsContainer component when fetching should render snapshot correctly 1`] = `ReactWrapper {}`;

exports[`CardsContainer component when initializes should render snapshot correctly 1`] = `ReactWrapper {}`;
